src/rust/engine/workunit_store/src/lib.rs

```diff
@@ -454,30 +462,52 @@ impl WorkunitStore {
       started.log_workunit_state()
     }
 
-    let sender = self.heavy_hitters_data.msg_tx.lock();
-    sender.send(StoreMsg::Started(started.clone())).unwrap();
+    if self.rendering_dynamic_ui {
+      let sender = self.heavy_hitters_data.msg_tx.lock();
+      sender.send(StoreMsg::Started(started.clone())).unwrap();
+    } else {
+      let mut inner = self.heavy_hitters_data.inner.lock();
+      HeavyHittersData::add_started_workunit_to_store(started.clone(), &mut inner);
```

Comment:
Rather than adding this to the store, could the caller hold onto a copy of the workunit until it was completed?

Reply:
I think that would make the logic overly complicated for no gain - the store already knows how to hold onto a started workunit, and then mark it as complete when the completed version of the workunit arrives.

Reply:
Hm. I think that I disagree. Writing to the store here would seem to break the abstraction that we enqueue things to be consumed, but otherwise don't acquire the lock on the store. It adds complexity because it violates that abstraction.

Holding onto the workunit between `let workunit = ...start_workunit(); ... complete_workunit(workunit)` on the other hand seems to fit into this model just fine, and would likely be less complex?

Reply:
The locks and queues are concerns internal to `WorkunitStore`; callers of `WorkunitStore::start_workunit()` and `complete_workunit` shouldn't care about what these methods are doing internally. This change would also require the signature of `start_workunit` and `complete_workunit` to effectively be different depending on whether the dynamic UI was active or not, if I'm understanding your suggestion correctly.

Reply:
> This change would also require the signature of start_workunit and complete_workunit to effectively be different depending on whether the dynamic UI was active or not, if I'm understanding your suggestion correctly.

I think that you could do this in all cases, and it would remove the dict/map lookup to figure out "which" workunit you are finishing. It might even be more efficient because of that. But also, any caller using `with_workunit` (which we should recommend!) wouldn't have to worry about it at all.

Reply:
And it would probably also fix https://github.com/pantsbuild/pants/issues/10249 ...?